processing/src/main/java/io/druid/query/extraction/namespace/ExtractionNamespaceFunctionFactory.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction.namespace;
+
+import com.google.common.base.Function;
+
+import java.util.Map;
+
+/**
+ *
+ */
+public interface ExtractionNamespaceFunctionFactory<T extends ExtractionNamespace>
+{
+  Function<String, String> build(T extractionNamespace);
+
+  Runnable getCachePopulator(T extractionNamespace, Map<String, String> swap);
```

Comment:
I am pretty sure that there will be people writing custom extensions implementing this. It will be great to have this interface and the methods documented with clear contract.

Now, if I did not look at all of the code, this is what I would guess.

Map build(T extractionNamespace): constructs mapping function for given namespace and returns it.

Runnable getCachePopulator(T extractionNamespace, Map swap):  constructs the mapping functions and puts those mappings inside "swap".

after looking at all the code,
It is expected that, the build(..) and cache populator Runnable share some state in common. cache populator is supposed to update that state and build is supposed to utilize that state.
Most of the implementations used NamespaceExtractionCacheManager.getCacheMap(namespace) as this "state". 

I am not sure how "swap" argument to getCachePopulator(..) making things better. It seems build(..) has to assume the fact that cache populator is updating some shared state. "swap" is shared between the two not by contract but by implementation detail and in fact, "swap" can be completely ignored as KafkaExtractionNamespaceFnFactory does.